File: myFit-api/src/main/java/com/example/fitme/module/user_metric/service/bodymetric/BodyMetricService.java
package com.example.fitme.module.user_metric.service.bodymetric;

import com.example.fitme.module.user_metric.dto.bodyMetric.BodyMetricRequest;
import com.example.fitme.module.user_metric.dto.bodyMetric.BodyMetricResponse;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for BodyMetric operations.
 * Defines CRUD operations for managing user body metrics.
 */
public interface BodyMetricService {

    /**
     * Create a new body metric record.
     * 
     * @param request The body metric data
     * @return Created body metric response
     */
    BodyMetricResponse create(BodyMetricRequest request);

    /**
     * Get all body metrics for a user, ordered by newest first.
     * 
     * @param userId The user ID
     * @return List of body metrics
     */
    List<BodyMetricResponse> getUserHistory(UUID userId);

    /**
     * Get the latest body metric for a user.
     * 
     * @param userId The user ID
     * @return Latest body metric
     */
    BodyMetricResponse getLatestByUserId(UUID userId);

    /**
     * Get a body metric by ID.
     * 
     * @param id The body metric ID
     * @return Body metric response
     */
    BodyMetricResponse getById(UUID id);

    /**
     * Update an existing body metric.
     * 
     * @param id The body metric ID
     * @param request The updated data
     * @return Updated body metric
     */
    BodyMetricResponse update(UUID id, BodyMetricRequest request);

    /**
     * Delete a body metric by ID.
     * 
     * @param id The body metric ID
     */
    void delete(UUID id);
}

File: myFit-FrontEnd/src/services/bodyMetricService.ts
import { api } from '../api/client';
import { ApiResponse, BodyMetricRequest, BodyMetricResponse } from '../types';

const BASE_PATH = '/api/body-metrics';


/**
 * Create new body metric
 * POST /api/body-metrics
 */
export const createBodyMetric = async (data: BodyMetricRequest): Promise<BodyMetricResponse> => {
    const response = await api.post<ApiResponse<BodyMetricResponse>>(BASE_PATH, data);

    if (response.data.code !== 1000) {
        throw new Error(response.data.message || 'Failed to create body metric');
    }

    return response.data.result;
};

/**
 * Get all body metrics for a user (ordered by newest first)
 * GET /api/body-metrics/user/{userId}
 */
export const getUserBodyMetrics = async (userId: string): Promise<BodyMetricResponse[]> => {
    const response = await api.get<ApiResponse<BodyMetricResponse[]>>(`${BASE_PATH}/user/${userId}`);

    if (response.data.code !== 1000) {
        throw new Error(response.data.message || 'Failed to fetch body metrics');
    }

    return response.data.result;
};

/**
 * Get latest body metric for a user
 * GET /api/body-metrics/user/{userId}/latest
 */
export const getLatestBodyMetric = async (userId: string): Promise<BodyMetricResponse> => {
    const response = await api.get<ApiResponse<BodyMetricResponse>>(`${BASE_PATH}/user/${userId}/latest`);

    if (response.data.code !== 1000) {
        throw new Error(response.data.message || 'Failed to fetch latest body metric');
    }

    return response.data.result;
};

/**
 * Get body metric by ID
 * GET /api/body-metrics/{id}
 */
export const getBodyMetricById = async (id: string): Promise<BodyMetricResponse> => {
    const response = await api.get<ApiResponse<BodyMetricResponse>>(`${BASE_PATH}/${id}`);

    if (response.data.code !== 1000) {
        throw new Error(response.data.message || 'Failed to fetch body metric');
    }

    return response.data.result;
};

/**
 * Update body metric
 * PUT /api/body-metrics/{id}
 */
export const updateBodyMetric = async (
    id: string,
    data: BodyMetricRequest
): Promise<BodyMetricResponse> => {
    const response = await api.put<ApiResponse<BodyMetricResponse>>(`${BASE_PATH}/${id}`, data);

    if (response.data.code !== 1000) {
        throw new Error(response.data.message || 'Failed to update body metric');
    }

    return response.data.result;
};

/**
 * Delete body metric
 * DELETE /api/body-metrics/{id}
 */
export const deleteBodyMetric = async (id: string): Promise<void> => {
    const response = await api.delete<ApiResponse<void>>(`${BASE_PATH}/${id}`);

    if (response.data.code !== 1000) {
        throw new Error(response.data.message || 'Failed to delete body metric');
    }
};
